File: exercises/csharp/TriviaGame/TriviaGame.Core/Game.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace TriviaGame.Core
{
  public class Game
  {
    public static readonly int NUMBER_OF_BOARD_SQUARES = 12;
    public static readonly int MINIMUM_PLAYERS = 2;
    public static readonly int MAX_QUESTIONS_PER_CATEGORY = 50;

    public bool IsPlayable => NumberOfPlayers >= MINIMUM_PLAYERS;
    public int NumberOfPlayers => _players.Count;

    private readonly string[] _categories = { "Pop", "Science", "Sports", "Rock" };
    private readonly List<string> _players = new List<string>();
    private readonly int[] _places = new int[6];
    private readonly int[] _purses = new int[6];
    private readonly bool[] _inPenaltyBox = new bool[6];
    private readonly Dictionary<string, List<string>> _questions = new Dictionary<string, List<string>>();

    private readonly LinkedList<string> _popQuestions = new LinkedList<string>();
    private readonly LinkedList<string> _scienceQuestions = new LinkedList<string>();
    private readonly LinkedList<string> _sportsQuestions = new LinkedList<string>();
    private readonly LinkedList<string> _rockQuestions = new LinkedList<string>();

    private bool _hasPlayerWon => CurrentPlayersGoldCoins != 6;
    private string _currentPlayerName => _players[_currentPlayer];
    private string _currentCategory => _categories[CurrentPlayerPlace % 4];

    private int _currentPlayer;
    private bool _isLeavingPenaltyBox;
    private bool IsCurrentPlayerInPenaltyBox
    {
      get => _inPenaltyBox[_currentPlayer];
      set => _inPenaltyBox[_currentPlayer] = value;
    }

    private int CurrentPlayerPlace
    {
      get => _places[_currentPlayer];
      set => _places[_currentPlayer] = value;
    }

    private int CurrentPlayersGoldCoins
    {
      get => _purses[_currentPlayer];
      set => _purses[_currentPlayer] = value;
    }

    public Game()
    {
      _questions = _categories.ToDictionary(
        category => category,
        category => Enumerable
          .Repeat($"{category} Question ", MAX_QUESTIONS_PER_CATEGORY)
          .Select((question, index) => question + index)
          .ToList());

      for (int index = 0; index < MAX_QUESTIONS_PER_CATEGORY; index++)
      {
        _popQuestions.AddLast("Pop Question " + index);
        _scienceQuestions.AddLast(("Science Question " + index));
        _sportsQuestions.AddLast(("Sports Question " + index));
        _rockQuestions.AddLast("Rock Question " + index);
      }
    }

    public bool AddPlayer(String playerName)
    {
      _players.Add(playerName);
      _places[NumberOfPlayers] = 0;
      _purses[NumberOfPlayers] = 0;
      _inPenaltyBox[NumberOfPlayers] = false;

      Console.WriteLine(playerName + " was added");
      Console.WriteLine("They are player number " + NumberOfPlayers);
      return true;
    }


    public void OnDieRoll(int roll)
    {
      Console.WriteLine(_currentPlayerName + " is the current player");
      Console.WriteLine("They have rolled a " + roll);

      if (IsCurrentPlayerInPenaltyBox)
      {
        var isRollOdd = roll % 2 != 0;
        Console.WriteLine(_currentPlayerName + $" is{(isRollOdd ? " " : " not ")}getting out of the penalty box");
        _isLeavingPenaltyBox = isRollOdd;

        if (!isRollOdd)
        {
          return;
        }
      }

      AdvancePlace(roll);
      PrintLocation();
      AskQuestion();
    }

    private void AdvancePlace(int roll)
    {
      CurrentPlayerPlace = (CurrentPlayerPlace + roll) % NUMBER_OF_BOARD_SQUARES;
    }

    private void MoveToNextPlayer()
    {
      _currentPlayer = (_currentPlayer + 1) % NumberOfPlayers;
    }

    private void AskQuestion()
    {
      var questions = _questions[_currentCategory];
      if (questions.Any())
      {
        System.Console.WriteLine(questions.First());
        questions.RemoveAt(0);
      }
    }

    public bool OnAnswer(bool isCorrect)
    {
      var winner = isCorrect
        ? OnCorrectAnswer()
        : OnIncorrectAnswer();

      MoveToNextPlayer();
      return winner;
    }

    public bool OnCorrectAnswer()
    {
      if (IsCurrentPlayerInPenaltyBox && !_isLeavingPenaltyBox)
      {
        return true;
      }

      var congratulationsMessage = IsCurrentPlayerInPenaltyBox
        ? "Answer was correct!!!!"
        : "Answer was corrent!!!!";

      Console.WriteLine(congratulationsMessage);
      CurrentPlayersGoldCoins++;
      PrintPlayerScore();
      var winner = _hasPlayerWon;

      return winner;
    }

    public bool OnIncorrectAnswer()
    {
      Console.WriteLine("Question was incorrectly answered");
      Console.WriteLine(_currentPlayerName + " was sent to the penalty box");
      IsCurrentPlayerInPenaltyBox = true;
      return true;
    }

    private void PrintPlayerScore()
    {
      Console.WriteLine(_currentPlayerName
                      + " now has "
                      + CurrentPlayersGoldCoins
                      + " Gold Coins.");
    }

    private void PrintLocation()
    {
      Console.WriteLine(_currentPlayerName
                + "'s new location is "
                + CurrentPlayerPlace);
      Console.WriteLine("The category is " + _currentCategory);
    }
  }
}

File: exercises/php/trivia/Game.php
<?php

namespace trivia;

function echoln($string)
{
	echo $string . "\n";
}

class Game
{
	var $players;
	var $places;
	var $board;
	var $purses;
	var $inPenaltyBox;

	var $popQuestions;
	var $scienceQuestions;
	var $sportsQuestions;
	var $rockQuestions;

	var $currentPlayer = 0;
	var $isGettingOutOfPenaltyBox;

	private const MAX_QUESTIONS_PER_CATEGORY = 50;

	function  __construct()
	{
		$this->board = new GameBoard();


		$this->players = array();
		$this->places = array(0);
		$this->purses  = array(0);
		$this->inPenaltyBox  = array(0);

		$this->popQuestions = array();
		$this->scienceQuestions = array();
		$this->sportsQuestions = array();
		$this->rockQuestions = array();

		for ($index = 0; $index < self::MAX_QUESTIONS_PER_CATEGORY; $index++) {
			array_push($this->popQuestions, "Pop Question " . $index);
			array_push($this->scienceQuestions, ("Science Question " . $index));
			array_push($this->sportsQuestions, ("Sports Question " . $index));
			array_push($this->rockQuestions, $this->createRockQuestion($index));
		}
	}

	function createRockQuestion($index)
	{
		return "Rock Question " . $index;
	}

	function isPlayable()
	{
		return ($this->getPlayerCount() >= 2);
	}

	function addPlayer($playerName)
	{
		array_push($this->players, $playerName);
		$this->places[$this->getPlayerCount()] = 0;
		$this->purses[$this->getPlayerCount()] = 0;
		$this->inPenaltyBox[$this->getPlayerCount()] = false;

		echoln($playerName . " was added");
		echoln("They are player number " . count($this->players));
		return true;
	}

	function getPlayerCount()
	{
		return count($this->players);
	}

	function  rollDice($roll)
	{
		echoln($this->players[$this->currentPlayer] . " is the current player");
		echoln("They have rolled a " . $roll);

		if ($this->inPenaltyBox[$this->currentPlayer]) {

			$this->isRollOddAndDoesOtherThings($roll);
		} else {
			$this->movePlaces($roll);
		}
	}

	function  askQuestion()
	{
		if ($this->board->getcurrentCategory($this->places[$this->currentPlayer]) == "Pop")
			echoln(array_shift($this->popQuestions));
		if ($this->board->getcurrentCategory($this->places[$this->currentPlayer]) == "Science")
			echoln(array_shift($this->scienceQuestions));
		if ($this->board->getcurrentCategory($this->places[$this->currentPlayer]) == "Sports")
			echoln(array_shift($this->sportsQuestions));
		if ($this->board->getcurrentCategory($this->places[$this->currentPlayer]) == "Rock")
			echoln(array_shift($this->rockQuestions));
	}




	function wasCorrectlyAnswered()
	{
		if ($this->inPenaltyBox[$this->currentPlayer]) {
			if ($this->isGettingOutOfPenaltyBox) {
				echoln("Answer was correct!!!!");
				$this->getAnswerOutcome('correct');

				$winner = $this->didPlayerWin();
				$this->currentPlayer++;
				if ($this->currentPlayer == count($this->players)) $this->currentPlayer = 0;

				return $winner;
			} else {
				$this->currentPlayer++;
				if ($this->currentPlayer == count($this->players)) $this->currentPlayer = 0;
				return true;
			}
		} else {

			echoln("Answer was corrent!!!!");
			$this->getAnswerOutcome('correct');

			$winner = $this->didPlayerWin();

			$this->nextPlayer();

			return $winner;
		}
	}

	function wasNotCorrectlyAnswered()
	{
		$this->getAnswerOutcome('incorrect');

		$this->inPenaltyBox[$this->currentPlayer] = true;

		$this->nextPlayer();

		return true;
	}


	function didPlayerWin()
	{
		return !($this->purses[$this->currentPlayer] == 6);
	}

	function nextPlayer()
	{
		$this->currentPlayer++;
		if ($this->currentPlayer == count($this->players)) $this->currentPlayer = 0;
	}

	function getAnswerOutcome($answer)
	{
		if ($answer === "correct") {
			$this->purses[$this->currentPlayer]++;
			echoln($this->players[$this->currentPlayer]
				. " now has "
				. $this->purses[$this->currentPlayer]
				. " Gold Coins.");
		} else {
			echoln("Question was incorrectly answered");
			echoln($this->players[$this->currentPlayer] . " was sent to the penalty box");
		}
	}

	function movePlaces($places)
	{

		$this->places[$this->currentPlayer] = $this->currentPlayerPlace()  + $places;
		if ($this->places[$this->currentPlayer] > 11) $this->places[$this->currentPlayer] = $this->currentPlayerPlace() - 12;

		echoln($this->players[$this->currentPlayer]
			. "'s new location is "
			. $this->currentPlayerPlace());
		echoln("The category is " . $this->board->getcurrentCategory($this->currentPlayerPlace()));

		$this->askQuestion();
	}

	function currentPlayerPlace()
	{
		return $this->places[$this->currentPlayer];
	}

	function isRollOddAndDoesOtherThings($roll)
	{
		if ($roll % 2 != 0) {
			$this->isGettingOutOfPenaltyBox = true;

			echoln($this->players[$this->currentPlayer] . " is getting out of the penalty box");
			$this->movePlaces($roll);
		} else {
			echoln($this->players[$this->currentPlayer] . " is not getting out of the penalty box");
			$this->isGettingOutOfPenaltyBox = false;
		}
	}
}
